Prose/ProseLib/Sources/ConversationFeature/Chat/Chat.swift

```diff
@@ -42,7 +47,13 @@ struct ProseCoreViewsMessage: Equatable, Encodable {
       jid: message.from.jidString,
       name: message.from.jidString
     )
-    self.reactions = message.reactions
+    self.reactions = message.reactions.reactions.map(Reaction.init(from:))
```

Comment:
Couldn't we adjust the `encode(to:)` method of `MessageReactions` or alternatively get rid of it?

Reply:
I wanted to keep a layer with types from prose-core-views. Nothing says `MessageReactions` is encoded in JSON as the web view needs it. And we shouldn't have to change our internal model if the web view API changes. I agree, right now we have the same structure, but nothing says it will stay so.

Reply:
By manually implementing the `encode(to:)` method we already have a layer of abstraction.

You can leave it like that, but the thing is that I'd like to have our models as close as possible to the (core-)view models. Over time - with conversations growing - mapping over each message (and now reactions) gets quite expensive since you have to allocate a new array each time and due to the reactive nature of our architecture this happens for every small change to the messages.

I've noticed that this becomes especially apparent when you're scrolling and inserting loaded older messages. The allocations can then be noticeable when the scrollview drops frames.

The JSON marshalling probably has the most overhead right now, I just want to give you my rationale.

Reply:
If you leave it like that please remove the `MessageReactions` conformance to `Encodable` though since its not used anymore.

Reply:
I get your point, and you're right we could change this.

My issue with it is just that `encode(to:)` is not linked by any mean to JSON. Only `JSONEncoder` and `JSONSerialization` are. Creating a custom `encode(to:)` just for the web view means we will never be able to encode for anything else (unless it uses the exact same format, which is highly improbable). If we could have some other function instead, which specifically targets the web view, I'm all for it!

Do you have an idea? What's better than mapping it to another struct? Using `[String: Any]` and `JSONSerialization`?

Reply:
The shape of the struct is not really the problem. The two problems that I'm seeing are:
1. `messages.map` is called for each change to `messages` and `Collection.map` allocates a new array for the mapped results.
2. JSON encoding is expensive but I wouldn't want to address that now. We can use proto- or flatbuffers at some point.

Overall I would like to treat the models from the ProseCoreTCA targets as our view models. And a requirement of our view models is that they need to be able to be serialized to JSON.

Sure if we'd want to store these for caching purposes or something and would want to have a different format there or make sure that it is stable we'd need to split the two serialization paths (for storage and for the webview).

This would most likely be done one layer deeper in the core lib though where we have separate models.

But that's getting too complicated for that format here already. I'll make a note so that we can address that briefly in our call.

Reply:
Should I open a new issue and fix this struct duplication tomorrow (or sometime soon) or should I handle it in this PR?

Reply:
"Sometime soon" sounds good to me :)

Reply:
I created #103. I will resolve this conversation and merge the PR now.